glib-macros/src/gclosure.rs

```diff
@@ -0,0 +1,220 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::utils::crate_ident_new;
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::{ToTokens, TokenStreamExt, quote};
+use syn::Token;
+
+#[derive(Clone, Copy, Debug)]
+enum CaptureKind {
+    Weak,
+    WeakAllowNone,
+    Strong,
+}
+
+struct Capture {
+    name: TokenStream,
+    alias: Option<syn::Ident>,
+    kind: CaptureKind,
+}
+
+impl Capture {
+    fn alias(&self) -> TokenStream {
+        if let Some(ref a) = self.alias {
+            a.to_token_stream()
+        } else {
+            self.name.to_token_stream()
+        }
+    }
+    fn outer_before_tokens(&self, crate_ident: &TokenStream) -> TokenStream {
+        let alias = self.alias();
+        let name = &self.name;
+        match self.kind {
+            CaptureKind::Weak | CaptureKind::WeakAllowNone => quote! {
+                let #alias = #crate_ident::clone::Downgrade::downgrade(&#name);
+            },
+            CaptureKind::Strong => quote! {
+                let #alias = #name.clone();
+            },
+        }
+    }
+
+    fn outer_after_tokens(&self, crate_ident: &TokenStream, closure_ident: &Ident) -> TokenStream {
+        let name = &self.name;
+        match self.kind {
+            CaptureKind::Weak => quote! {
+                #crate_ident::object::Object::watch_closure(&#name, &#closure_ident);
+            },
+            _ => Default::default(),
+        }
+    }
+
+    fn inner_before_tokens(&self, crate_ident: &TokenStream) -> TokenStream {
+        let alias = self.alias();
+        match self.kind {
+            CaptureKind::Weak => {
+                let err_msg = format!("failed to upgrade `{}`", alias.to_string());
+                quote! {
+                    let #alias = match #crate_ident::clone::Upgrade::upgrade(&#alias) {
+                        Some(val) => val,
+                        None => panic!(#err_msg)
+                    };
+                }
+            },
+            CaptureKind::WeakAllowNone => quote! {
+                let #alias = #crate_ident::clone::Upgrade::upgrade(&#alias);
```

Comment:
Should there maybe be a variant that automatically invalidates the closure (i.e. doesn't actually call the body of the closure) if upgrading any of the weak references fails? That seems generally useful and then you don't need to implement that part in the gtk4-rs PR.

Reply:
Calling `g_closure_invalidate()` from here seems a bit tricky as you somehow need to channel the `GClosure*` into the closure itself (needs some `Arc<Mutex<_>>` construct in the general case), but simply not calling the closure body in this case seems a good start already.

Reply:
I think this should already be handled by calling `watch_closure`. That was my major motivation for doing this. :) See the code in `outer_after_tokens`. Maybe the documentation is not as clear as it could be?

Reply:
I think I just missed that part in the documentation :) Sorry

Reply:
This is of course not threadsafe: the closure might just be called, then is invalidated, then you try upgrading the weak reference but the object is just gone.

So it would probably be good to also handle this case here instead of panicking

Reply:
That should never happen, `watch_closure` will ensure the object has at least one strong reference while the closure is being called. See [here](https://gitlab.gnome.org/GNOME/glib/-/blob/b4b5b3d029c1b5f8c859657bc269c39f276b3c68/gobject/gobject.c#L4426). And you can see this in the unit test for `gclosure!`, the ref count of a `@weak` capture always has a ref count of at least 3.

Reply:
Actually, now that I try this, using `@weak` on multiple objects doesn't work, the `GClosure` is limited to watching only one object. So that will have to be removed.

Reply:
How do you want to handle that? By just not calling `watch()` and simply not calling the closure when object can't be upgraded? That however won't invalidate the closure and disconnect the signal handler, so we somehow need to also invalidate it.

Reply:
I was thinking to have the macro abort if more than one `@weak` is used. To support more would probably need to be changed in gobject by adding an internal mutex to `GClosure`.

Reply:
That also works, yes

Reply:
The macro now aborts when more than one object is watched